pkg/gardenlet/bootstrap/bootstrap_test.go

```diff
@@ -31,39 +35,129 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	fakeclient "sigs.k8s.io/controller-runtime/pkg/client/fake"
 )
 
 var _ = Describe("Bootstrap", func() {
-	Describe("BuildBootstrapperName", func() {
-		It("should return the correct name", func() {
-			name := "foo"
-			result := BuildBootstrapperName(name)
-			Expect(result).To(Equal(fmt.Sprintf("%s:%s", GardenerSeedBootstrapper, name)))
-		})
+	var (
+		ctrl       *gomock.Controller
+		c          *mockclient.MockClient
+		ctx        = context.TODO()
+		testLogger = logger.NewFieldLogger(logger.NewLogger("info"), "test", "cert")
+	)
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		c = mockclient.NewMockClient(ctrl)
 	})
 
-	Describe("#DeleteBootstrapAuth", func() {
-		var (
-			ctrl *gomock.Controller
-			c    *mockclient.MockClient
+	AfterEach(func() {
+		ctrl.Finish()
+	})
 
-			ctx     = context.TODO()
-			csrName = "csr-name"
-			csrKey  = kutil.Key(csrName)
+	Describe("#BootstrapKubeconfig", func() {
+		var (
+			seedName = "test"
+			seed     gardencorev1beta1.Seed
+
+			seedClient            client.Client
+			bootstrapClientConfig *rest.Config
+
+			gardenClientConnection *config.GardenClientConnection
+			seedConfig             config.SeedConfig
+
+			approvedCSR = certificatesv1beta1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "watched-csr",
+				},
+				Status: certificatesv1beta1.CertificateSigningRequestStatus{
+					Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{
+						{
+							Type: certificatesv1beta1.CertificateApproved,
+						},
+					},
+					Certificate: []byte("my-cert"),
+				},
+			}
+
+			deniedCSR = certificatesv1beta1.CertificateSigningRequest{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: "watched-csr",
+				},
+				Status: certificatesv1beta1.CertificateSigningRequestStatus{
+					Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{
+						{
+							Type: certificatesv1beta1.CertificateDenied,
+						},
+					},
+				},
+			}
 		)
 
 		BeforeEach(func() {
-			ctrl = gomock.NewController(GinkgoT())
-			c = mockclient.NewMockClient(ctrl)
+			secretReference := corev1.SecretReference{
+				Name:      "bootstrap-kubeconfig",
+				Namespace: "garden",
+			}
+
+			// gardenClientConnection with required bootstrap secret kubeconfig secret
+			// in a non-test environment we would use two different secrets
+			gardenClientConnection = &config.GardenClientConnection{
+				BootstrapKubeconfig: &secretReference,
+				KubeconfigSecret:    &secretReference,
+			}
+
+			// seedConfig containing the seed name
+			seed = gardencorev1beta1.Seed{
+				ObjectMeta: metav1.ObjectMeta{Name: seedName},
+			}
+			seedConfig = config.SeedConfig{Seed: seed}
+
+			// create mock seed client
+			s := runtime.NewScheme()
+			Expect(corev1.AddToScheme(s)).ToNot(HaveOccurred())
+			Expect(gardencorescheme.AddToScheme(s)).ToNot(HaveOccurred())
+			seedClient = fakeclient.NewFakeClientWithScheme(s, &seed)
```

Comment:
Why do you add the `Seed` object to the fake client? There's no `Seed` object on the Seed cluster itself...
```suggestion
			seedClient = fakeclient.NewFakeClientWithScheme(scheme.Scheme)
```

Reply:
Correct, I do not need to add the seed object - thanks for spotting. 
But I still need the corev1 scheme [otherwise the secret in the seed cannot be updated](https://github.com/gardener/gardener/blob/67833d9f9bbfd0b9a8ddd0cfffec6a455ead2586/pkg/gardenlet/bootstrap/certificate/certificate_rotation.go#L221-L220)

Reply:
Yes. That's why I suggested to just use the default client-go scheme (`scheme.Scheme`) as it already contains `corev1` ([ref](https://github.com/kubernetes/client-go/blob/36233866f1c7c0ad3bdac1fc466cb5de3746cfa2/kubernetes/scheme/register.go#L93) and not bother with constructing a new scheme.
Anyways, it's fine that way, too.